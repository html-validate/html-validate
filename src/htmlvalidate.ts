import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
