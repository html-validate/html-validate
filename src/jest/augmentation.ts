/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
