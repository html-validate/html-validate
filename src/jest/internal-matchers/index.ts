export { toBeToken } from "./to-be-token";
