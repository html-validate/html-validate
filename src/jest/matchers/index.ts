export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
