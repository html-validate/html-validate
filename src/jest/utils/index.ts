export { diff } from "./diff";
export { type MaybeAsyncCallback, diverge } from "./diverge";
export { flattenMessages } from "./flatten-messages";
export { isThenable } from "./is-thenable";
export { type MatcherContext } from "./matcher-context";
export { type MatcherExpect } from "./matcher-expect";
export { type MatcherResult } from "./matcher-result";
export { stripAnsi } from "./strip-ansi";
