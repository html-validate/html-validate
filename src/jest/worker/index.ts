export { type ValidateStringFn } from "./worker";
export { createSyncFn } from "./create-sync-fn";
export { default as workerPath } from "./worker.ts?worker&url";
