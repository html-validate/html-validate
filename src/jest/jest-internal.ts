import "./jest";

import { Location } from "../context";
import { TokenType } from "../lexer";
import { toBeToken } from "./internal-matchers";

interface TokenMatcher {
	type: TokenType;
	location?: Partial<Location>;
	data?: any;
}

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-unused-vars */
		interface Matchers<R, T = {}> {
			toBeToken(expected: TokenMatcher): R;
		}
	}
}

expect.extend({
	toBeToken,
});
