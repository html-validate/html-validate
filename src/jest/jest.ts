import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
