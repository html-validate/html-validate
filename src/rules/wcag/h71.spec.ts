import { HtmlValidate } from "../../htmlvalidate";
import "../../jest";

describe("wcag/h71", () => {
	let htmlvalidate: HtmlValidate;

	beforeAll(() => {
		htmlvalidate = new HtmlValidate({
			elements: [
				"html5",
				{
					custom: {
						inherit: "fieldset",
					},
				},
			],
			rules: { "wcag/h71": "error" },
		});
	});

	it("should report error when <fieldset> is missing <legend>", async () => {
		expect.assertions(2);
		const report = await htmlvalidate.validateString("<fieldset></fieldset>");
		expect(report).toBeInvalid();
		expect(report).toHaveError("wcag/h71", "<fieldset> must have a <legend> as the first child");
	});

	it("should report error when custom element inherits from <fieldset>", async () => {
		expect.assertions(2);
		const report = await htmlvalidate.validateString("<custom></custom>");
		expect(report).toBeInvalid();
		expect(report).toHaveError("wcag/h71", "<custom> must have a <legend> as the first child");
	});

	it("should not report when <fieldset> have <legend>", async () => {
		expect.assertions(1);
		const report = await htmlvalidate.validateString("<fieldset><legend>foo</legend></fieldset>");
		expect(report).toBeValid();
	});

	it("should not report when <fieldset> have multiple <legend>", async () => {
		expect.assertions(1);
		const report = await htmlvalidate.validateString(
			"<fieldset><legend>foo</legend><legend>bar</legend></fieldset>",
		);
		expect(report).toBeValid();
	});

	it("should not report when <fieldset> have out-of-order <legend>", async () => {
		expect.assertions(1);
		const report = await htmlvalidate.validateString(
			"<fieldset><div>foo</div><legend>bar</legend></fieldset>",
		);
		expect(report).toBeValid();
	});

	it("should contain documentation", async () => {
		expect.assertions(1);
		htmlvalidate = new HtmlValidate({
			rules: { "wcag/h71": "error" },
		});
		const docs = await htmlvalidate.getContextualDocumentation({ ruleId: "wcag/h71" });
		expect(docs).toMatchSnapshot();
	});
});
