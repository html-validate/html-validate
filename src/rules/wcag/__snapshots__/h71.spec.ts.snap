// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wcag/h71 should contain documentation 1`] = `
{
  "description": "H71: Providing a description for groups of form controls using fieldset and legend elements",
  "url": "https://html-validate.org/rules/wcag/h71.html",
}
`;
